### PR TITLE
Decline Node feature

### LINKED ISSUE #27: On the Nodes page, I should be able to hide addresses I'm not going to approve - 8SP
In GitLab by @kevinbader on Aug 9, 2018, 14:54

### PR DESCRIPTION
**Pull request by georgimld**
### Checklist

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/openkfw/TruBudget/blob/master/CONTRIBUTING.md).
- [x] I fixed all necessary PR warnings
- [x] The commit history is clean
- [x] The E2E tests are passing
- [x] If possible, the issue has been divided into more subtasks
- [x] I did a self review before requesting a review from another team member

#### API:
**sending the decline request**: 
- [x] new endpoint that receives an object (node) that has the address and organization of the node that should be declined
- [x] save a new event on the `nodes`stream, that contains the address and orga of the node to be declined and the address and orga of the decliner (`/service/domain/network/node_declined.ts`)
- [x] Also check that a node with this address and orga has already been registered (`network/controller/declineNode.ts`)
- [x] check that the node to be declined has not been declined or approved yet by someone from the same orga as the decliner (also `network/controller/declineNode.ts`)
- [x] when approving a node also check that the node has not been declined yet by someone from the same orga as the approver (`network/controller/approveNewNodeForExistingOrganization.ts`)

**eventsourcing and listing nodes**:
- [x] when parsing the events in the nodes stream, a new entry is being saved to the NodeInfo of each node, where all decliners are saved with their address and orga (`Nodes.ts`). This new entry in the NodeInfo is used when checking if a node has already been declined or not (the checks mentioned above)
- [x] in the `list` function (`network.list` endpoint), the NodeInfo is being converted in the NodeInfoDto. Here, the decliners are saved on the same level of the response object as the approvers, in a similar array. 

#### Frontend: 

- [x] added a new button, action and saga to send request to api
- [x] when listing the requests, filter out the nodes that have already been declined by the organization of the logged in user

#### Unit-tests

- [x] Test where an event is saved successfully
- [x] Test when the address is not valid - expect an error to be thrown
- [x] Test when the node was already declined by the organization of the decliner
- [x] Test when a node was already approved (already has connect permissions)

**Note:** Tests where implemented using [sinon stubs](https://sinonjs.org/releases/latest/stubs/)

#### For reviewers:
- the new docker compose file `docker-compose/multi/docker-compose-multi-node.yml` will start the project with multiple slave nodes attempting to connect to the network. Testing the feature can then be done via postman or through the frontend. 
- using the [multichain-cli](https://github.com/openkfw/TruBudget/blob/master/doc/tutorials/multichain-cli/multichain-cli.md) you can check out the `nodes` stream to see what new events were saved (`liststreamitems nodes`)

#### TL;DR: new feature, awesome stuff

Closes #27

### CODE REVIEW

**Comment by georgimld** [comment]
> Warnings
> ⚠️	It looks like an ID was added. Please make sure it is not a secret.
> ⚠️	There were changes in the frontend, but no E2E-test was added or modified!
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [4dd61d4](https://github.com/openkfw/TruBudget/commit/4dd61d4a2a90a3dc5b46b6e5a782e8e5a8970522)

**Review by Stezido — CHANGES_REQUESTED** [review]

**Review by Stezido — APPROVED** [review]
